src/lang/std/artifactGraph.ts

```diff
@@ -0,0 +1,645 @@
+import { Models } from '@kittycad/lib'
+import { getNodePathFromSourceRange } from 'lang/queryAst'
+import { err } from 'lib/trap'
+
+interface CommonCommandProperties {
+  range: SourceRange
+  pathToNode: PathToNode
+}
+
+export interface _PlaneArtifact {
+  type: 'plane'
+  pathIds: Array<string>
+  codeRef: CommonCommandProperties
+}
+export interface PlaneArtifact {
+  type: 'plane'
+  paths: Array<_PathArtifact>
+  codeRef: CommonCommandProperties
+}
+
+export interface _PathArtifact {
+  type: 'path'
+  planeId: string
+  segIds: Array<string>
+  extrusionId: string
+  solid2dId?: string
+  codeRef: CommonCommandProperties
+}
+
+interface _solid2D {
+  type: 'solid2D'
+  pathId: string
+}
+export interface PathArtifact {
+  type: 'path'
+  plane: _PlaneArtifact | _WallArtifact
+  segs: Array<_SegmentArtifact>
+  extrusion: _ExtrusionArtifact
+  codeRef: CommonCommandProperties
+}
+
+interface _SegmentArtifact {
+  type: 'segment'
+  pathId: string
+  surfId: string
+  edgeIds: Array<string>
+  edgeCutId?: string
+  codeRef: CommonCommandProperties
+}
+interface SegmentArtifact {
+  type: 'segment'
+  path: _PathArtifact
+  surf: _WallArtifact
+  edges: Array<_ExtrudeEdge>
+  edgeCut?: _EdgeCut
+  codeRef: CommonCommandProperties
+}
+
+interface _ExtrusionArtifact {
+  type: 'extrusion'
+  pathId: string
+  surfIds: Array<string>
+  edgeIds: Array<string>
+  codeRef: CommonCommandProperties
+}
+interface ExtrusionArtifact {
+  type: 'extrusion'
+  pathId: string
+  surfs: Array<_WallArtifact | _CapArtifact>
+  edges: Array<string>
+  codeRef: CommonCommandProperties
+}
+
+interface _WallArtifact {
+  type: 'wall'
+  segId: string
+  edgeCutEdgeIds: Array<string>
+  extrusionId: string
+  pathIds: Array<string>
+}
+interface _CapArtifact {
+  type: 'cap'
+  subType: 'start' | 'end'
+  edgeCutEdgeIds: Array<string>
+  extrusionId: string
+  pathIds: Array<string>
+}
+
+interface _ExtrudeEdge {
+  type: 'extrudeEdge'
+  segId: string
+  extrusionId: string
+  edgeId: string
+}
+
+/** A edgeCut is a more generic term for both fillet or chamfer */
+interface _EdgeCut {
+  type: 'edgeCut'
+  subType: 'fillet' | 'chamfer'
+  consumedEdgeId: string
+  edgeIds: Array<string>
+  surfId: string
+  codeRef: CommonCommandProperties
+}
+
+interface EdgeCutEdge {
+  type: 'edgeCutEdge'
+  edgeCutId: string
+  surfId: string
+}
+
+export type Artifact =
+  | _PlaneArtifact
+  | _PathArtifact
+  | _SegmentArtifact
+  | _ExtrusionArtifact
+  | _WallArtifact
+  | _CapArtifact
+  | _ExtrudeEdge
+  | _EdgeCut
+  | EdgeCutEdge
+  | _solid2D
+
+export type ArtifactGraph = Map<string, Artifact>
+
+export type EngineCommand = Models['WebSocketRequest_type']
+
+type OkWebSocketResponseData = Models['OkWebSocketResponseData_type']
+
+export interface ResponseMap {
+  [commandId: string]: OkWebSocketResponseData
+}
+export interface OrderedCommand {
+  command: EngineCommand
+  range: SourceRange
+}
+
+export function createArtifactGraph({
```

Comment:
wanted to keep this function implementation pretty simple, so it basically calls `getArtifactsToUpdate` in a loop, and uses that output to mutate the map.